xlators/storage/posix/src/posix-aio.c

```diff
@@ -391,14 +388,133 @@ posix_aio_writev(call_frame_t *frame, xlator_t *this, fd_t *fd,
 err:
     STACK_UNWIND_STRICT(writev, frame, -1, op_errno, 0, 0, 0);
 
-    if (paiocb) {
-        if (paiocb->iobref)
-            iobref_unref(paiocb->iobref);
-        if (paiocb->fd)
-            fd_unref(paiocb->fd);
-        GF_FREE(paiocb);
+    posix_aio_cb_fini(paiocb);
+
+    return 0;
+}
+
+int
+posix_aio_fsync_complete(struct posix_aio_cb *paiocb, int res, int res2)
+{
+    call_frame_t *frame = NULL;
+    xlator_t *this = NULL;
+    struct iatt prebuf = {
+        0,
+    };
+    struct iatt postbuf = {
+        0,
+    };
+    int _fd = -1;
+    int op_ret = -1;
+    int op_errno = 0;
+    int ret = 0;
+    fd_t *fd = NULL;
+
+    GF_ASSERT(paiocb);
+
+    frame = paiocb->frame;
+    this = frame->this;
+    prebuf = paiocb->prebuf;
+    fd = paiocb->fd;
+    _fd = paiocb->_fd;
+
+    if (res < 0) {
+        op_ret = -1;
+        op_errno = -res;
+        gf_msg(this->name, GF_LOG_ERROR, op_errno, P_MSG_FSYNC_FAILED,
+               "%s(async) failed fd=%d (%d)",
+               (paiocb->iocb.aio_lio_opcode == IO_CMD_FDSYNC ? "fdatasync"
+                                                             : "fsync"),
+               _fd, res);
+        goto out;
+    }
+
+    ret = posix_fdstat(this, fd->inode, _fd, &postbuf);
+    if (ret != 0) {
+        op_ret = -1;
+        op_errno = errno;
+        gf_msg(this->name, GF_LOG_ERROR, op_errno, P_MSG_FSTAT_FAILED,
+               "fstat failed on fd=%d", _fd);
+        goto out;
+    }
+
+    op_ret = res;
+    op_errno = 0;
+
+out:
+    STACK_UNWIND_STRICT(fsync, frame, op_ret, op_errno, &prebuf, &postbuf,
+                        NULL);
+
+    posix_aio_cb_fini(paiocb);
+
+    return 0;
+}
+
+int
+posix_aio_fsync(call_frame_t *frame, xlator_t *this, fd_t *fd, int32_t datasync,
+                dict_t *xdata)
+{
+    int32_t op_errno = EINVAL;
+    int _fd = -1, ret = -1;
+    struct posix_fd *pfd = NULL;
+    struct posix_aio_cb *paiocb = NULL;
+    struct posix_private *priv = NULL;
+    struct iocb *iocb = NULL;
+
+    VALIDATE_OR_GOTO(frame, err);
+    VALIDATE_OR_GOTO(this, err);
+    VALIDATE_OR_GOTO(fd, err);
+
+    priv = this->private;
+
+    ret = posix_fd_ctx_get(fd, this, &pfd, &op_errno);
+    if (ret < 0) {
+        gf_msg(this->name, GF_LOG_WARNING, op_errno, P_MSG_PFD_NULL,
+               "pfd is NULL from fd=%p", fd);
+        goto err;
+    }
+
+    _fd = pfd->fd;
+
+    paiocb = posix_aio_cb_init(frame, fd, _fd, GF_FOP_FSYNC);
+    if (!paiocb) {
+        op_errno = ENOMEM;
+        goto err;
+    }
+
+    paiocb->iocb.aio_lio_opcode = datasync ? IO_CMD_FDSYNC : IO_CMD_FSYNC;
+
+    iocb = &paiocb->iocb;
+
+    ret = posix_fdstat(this, fd->inode, _fd, &paiocb->prebuf);
+    if (ret != 0) {
+        op_errno = errno;
+        gf_msg(this->name, GF_LOG_ERROR, op_errno, P_MSG_FSTAT_FAILED,
+               "fstat failed on fd=%p", fd);
+        goto err;
+    }
+
+    LOCK(&fd->lock);
+    {
+        ret = io_submit(priv->ctxp, 1, &iocb);
+    }
+    UNLOCK(&fd->lock);
```

Comment:
Doing a system call inside a mutex doesn't seem a good idea unless there's some very important reason that cannot be solved in any other way. I don't see any reason here. Is there anything that can justify the lock ?

Reply:
Nice catch. If so, io_submit() should be called without holding per-fd locks in posix_aio_readv() and posix_aio_writev() as well.

Reply:
I don't like it, but I don't think that the lock around `io_submit()` can be easily removed for reads and writes. I'm not sure about the reasons, but it seems that the `O_DIRECT` flags is being dynamically changed on a per-request basis. I guess the lock is there to ensure that request is submitted with the desired `O_DIRECT`. If we move the `io_submit()` outside the lock, the other threads could change it before submitting the request.

I think `FSYNC` doesn't depend on `O_DIRECT`, so it can be submitted without taking the lock.

However, are we sure that the flags are checked and copied before returning from `io_submit()` ? otherwise the lock is useless to prevent asynchronous changes of the `O_DIRECT` flag. Anyway this is something for another patch.

Reply:
> are we sure that the flags are checked and copied before returning from io_submit() ? 

I've taken a look through the kernel's fs/aio.c and it seems the answer is yes - O_DIRECT is checked and copied very early when the data comes from userspace is verified and converted to internal kernel data structures.